Support nunit3

jhamm opened the pull request:
Here are the changes to support nunit3. This should fix #98.

The structure has some similarities as the XUnit Property.

If the result is exhausted, the test will be marked as Failed.

The failure messages are based on the recommendations on [Running Tests](https://fscheck.github.io/FsCheck/RunningTests.html) and the FsCheck XUnit Property project.

In the two example Nunit Test projects, there are now a total of 4 expected failing tests that end with _shouldFail. One of the tests is the exhausted test.

Here is the command that was used for testing this PR from the FsCheck directory:
`.\packages\NUnit.Console\tools\nunit-console.exe examples\FsCheck.NUnit.Examples\bin\debug\FsCheck.NUnit.Examples.dll examples\FsCheck.NUnit.CSharpExamples\bin\debug\FsCheck.NUnit.CSharpExamples.dll`

jhamm commented:
I will take a look at fixing the nuget side of this over the weekend.

kurtschelfthout commented on FsCheck.sln:
```diff
@@ -38,8 +38,6 @@ Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "FsCheck.Test", "tests\FsChe
 Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "FsCheck.NUnit", "src\FsCheck.NUnit\FsCheck.NUnit.fsproj", "{6B1B9FF5-8E6D-4461-9D77-CD1619403684}"
 EndProject
-Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "FsCheck.NUnit.Addin", "src\FsCheck.NUnit.Addin\FsCheck.NUnit.Addin.fsproj", "{BF24DE38-C529-4638-BE50-A937EA3E6876}"
-EndProject
```
Yay.

kurtschelfthout commented on examples/FsCheck.NUnit.CSharpExamples/FsCheck.NUnit.CSharpExamples.csproj:
```diff
@@ -35,6 +35,7 @@
     <Prefer32Bit>false</Prefer32Bit>
   </PropertyGroup>
   <ItemGroup>
+    <Reference Include="FSharp.Core, Version=4.3.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL" />
```
This is a dup.

kurtschelfthout commented on examples/FsCheck.NUnit.CSharpExamples/FsCheckNunitAddin.cs:
```diff
@@ -1,16 +1,15 @@
+    //        return true;
+    //    }
+    //}
 }
```
Please remove the empty file.

kurtschelfthout commented on examples/FsCheck.XUnit.CSharpExamples/FsCheck.XUnit.CSharpExamples.csproj:
```diff
@@ -67,7 +67,7 @@
   <Import Project="..\..\.paket\paket.targets" />
   <Choose>
-    <When Condition="($(TargetFrameworkIdentifier) == '.NETFramework' And ($(TargetFrameworkVersion) == 'v2.0' Or $(TargetFrameworkVersion) == 'v3.0' Or $(TargetFrameworkVersion) == 'v3.5' Or $(TargetFrameworkVersion) == 'v4.0' Or $(TargetFrameworkVersion) == 'v4.5' Or $(TargetFrameworkVersion) == 'v4.5.1' Or $(TargetFrameworkVersion) == 'v4.5.2' Or $(TargetFrameworkVersion) == 'v4.5.3' Or $(Targ ...
+    <When Condition="$(TargetFrameworkIdentifier) == '.NETFramework' And ($(TargetFrameworkVersion) == 'v2.0' Or $(TargetFrameworkVersion) == 'v3.0' Or $(TargetFrameworkVersion) == 'v3.5' Or $(TargetFrameworkVersion) == 'v4.0' Or $(TargetFrameworkVersion) == 'v4.5' Or $(TargetFrameworkVersion) == 'v4.5.1' Or $(TargetFrameworkVersion) == 'v4.5.2' Or $(TargetFrameworkVersion) == 'v4.5.3' Or $(Targe ...
```
Is this change intentional? Doesn't seem related to NUnit.

kurtschelfthout commented on paket.lock:
```diff
@@ -1,17 +1,18 @@
-    FSharp.Formatting (2.9.10)
+    FAKE (4.3.1)
+    FSharp.Compiler.Service (1.4.0.1)
+    FSharp.Formatting (2.10.0)
```
Can you roll this back, I'll update dependencies on master at some point but prefer not to conflate.

kurtschelfthout commented on paket.lock:
```diff
@@ -1,17 +1,18 @@
+    NUnit (3.0.0-beta-4)
+    NUnit.Console (3.0.0-beta-4)
+    NUnitTestAdapter (2.0.0)
+    SourceLink.Fake (1.1.0)
```
SourceLink, ditto.

kurtschelfthout commented on src/FsCheck.NUnit/FsCheckPropertyAttribute.fs:
```diff
@@ -39,3 +77,101 @@ type PropertyAttribute() =
+            testResult.SetResult(new ResultState(TestStatus.Failed, msg))
+        | TestResult.False (testdata, originalArgs, shrunkArgs, outcome, seed) ->
+            let msg = sprintf "%s" (Runner.onFinishedToString "" testRunner.Result)
+            testResult.SetResult(new ResultState(TestStatus.Failed, msg))
```
I'll take your word for it on this one... :)

kurtschelfthout commented on src/FsCheck.Xunit/FsCheck.Xunit.fsproj:
```diff
@@ -87,7 +87,7 @@
   <Import Project="..\..\.paket\paket.targets" />
   <Choose>
-    <When Condition="($(TargetFrameworkIdentifier) == '.NETFramework' And ($(TargetFrameworkVersion) == 'v2.0' Or $(TargetFrameworkVersion) == 'v3.0' Or $(TargetFrameworkVersion) == 'v3.5' Or $(TargetFrameworkVersion) == 'v4.0' Or $(TargetFrameworkVersion) == 'v4.5' Or $(TargetFrameworkVersion) == 'v4.5.1' Or $(TargetFrameworkVersion) == 'v4.5.2' Or $(TargetFrameworkVersion) == 'v4.5.3' Or $(Targ ...
+    <When Condition="$(TargetFrameworkIdentifier) == '.NETFramework' And ($(TargetFrameworkVersion) == 'v2.0' Or $(TargetFrameworkVersion) == 'v3.0' Or $(TargetFrameworkVersion) == 'v3.5' Or $(TargetFrameworkVersion) == 'v4.0' Or $(TargetFrameworkVersion) == 'v4.5' Or $(TargetFrameworkVersion) == 'v4.5.1' Or $(TargetFrameworkVersion) == 'v4.5.2' Or $(TargetFrameworkVersion) == 'v4.5.3' Or $(Targe ...
```
Unclear where this change comes from...?

kurtschelfthout commented on tests/FsCheck.Test/FsCheck.Test.fsproj:
```diff
@@ -114,18 +120,18 @@
+    <When Condition="($(TargetFrameworkIdentifier) == 'WindowsPhone' And ($(TargetFrameworkVersion) == 'v8.0' Or $(TargetFrameworkVersion) == 'v8.1')) Or ($(TargetFrameworkIdentifier) == 'MonoAndroid') Or ($(TargetFrameworkIdentifier) == 'MonoTouch') Or ($(TargetFrameworkIdentifier) == 'Xamarin.iOS') Or ($(TargetFrameworkIdentifier) == 'Xamarin.Mac') Or ($(TargetFrameworkProfile) == 'Profile7') O ...
       <PropertyGroup>
         <__paket__xunit_runner_visualstudio_props>portable-net45+aspnetcore50+win+wpa81+wp80+monotouch+monoandroid\xunit.runner.visualstudio</__paket__xunit_runner_visualstudio_props>
       </PropertyGroup>
```
Is this a re-ordering of some kind?

kurtschelfthout commented on tests/FsCheck.Test/FsCheck.Test.fsproj:
```diff
@@ -96,8 +96,14 @@
+      <xunitrunner Include="$(ProjectDir)..\..\packages\xunit.runner.visualstudio\**\*.*" />
+    </ItemGroup>
+    <Copy SourceFiles="@(xunitrunner)" DestinationFolder="$(ProjectDir)..\..\packages\xunit.runner.visualstudio.2.0.0\%(RecursiveDir)" SkipUnchangedFiles="true" />
+  </Target>
```
This target should no longer be necessary with the version of paket; it looks like this is a merge problem.

kurtschelfthout commented:
Very nice, loved to see FsCheck.NUnit.Addin go. Bunch of comments inline.

jhamm commented:
Thanks for the feedback! I will work on the changes over the weekend.

jhamm commented:
The changes you have suggested have been applied. The diff looks much cleaner now. Thanks!

There were a few other small things that were modified as well.

Updated the Documentation to remove the note about the Temporary fix to get NUnit.Addin working for your test.
The Documentation_ja.md still needs this modification.

Removed references to the FsCheck.NUnit.Addin from the mono solution, build.fsx, and FsCheck.Nunit.Examples project

Modified FsCheck.NUnit.nuspec:
to remove install.ps1 and FsCheckAddin.fs because they are no longer needed.
to not reference FsCheck.NUnit.Addin
remove NUnit.Runners from the dependency list

kurtschelfthout commented:
This is merged on branch `nunit3`. Thanks a bunch for all of this! I'll monitor the AppVeyor build and update #140 with the build number so people can test.
